extensions/arrow/src/main/java/io/deephaven/extensions/arrow/ArrowWrapperTools.java

```diff
@@ -0,0 +1,340 @@
+            } else if (vector instanceof Decimal256Vector) {
+                return new ArrowObjectColumnSource<>(BigDecimal.class, highBit, field, arrowHelper);
+            } else if (vector instanceof DateMilliVector) {
+                return new ArrowObjectColumnSource<>(LocalDateTime.class, highBit, field, arrowHelper);
+            } else if (vector instanceof TimeMilliVector) {
+                return new ArrowLocalTimeColumnSource(highBit, field, arrowHelper);
+            } else if (vector instanceof TimeStampVector) {
+                return new ArrowDateTimeColumnSource(highBit, field, arrowHelper);
+            }
+            throw new TableDataException(
+                    String.format("Unsupported vector: name: %s type: %s", vector.getName(), vector.getMinorType()));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        } finally {
+            arrowHelper.addToPool(context);
+        }
+    }
+
+    private static class ArrowWrappedTable extends QueryTable {
+        private final ArrowWrapperTools.Helper helper;
+
+        public ArrowWrappedTable(
+                final TrackingRowSet rowSet,
+                final Map<String, ? extends ColumnSource<?>> columns,
+                final ArrowWrapperTools.Helper helper) {
+            super(rowSet, columns);
+            this.helper = helper;
+        }
+
+        @Override
+        public void close() {
+            super.close();
+            this.helper.close();
+        }
+    }
+
+    private static final class SharingKey extends SharedContext.ExactReferenceSharingKey<Shareable> {
+        public SharingKey(@NotNull final Helper arrowHelper) {
+            super(arrowHelper);
+        }
+    }
+
+    public static class FillContext implements ColumnSource.FillContext {
+        private final boolean shared;
+        private final Shareable shareable;
+        private final Helper helper;
+
+        public FillContext(final Helper helper, final SharedContext sharedContext) {
+            this.helper = helper;
+            this.shared = sharedContext != null;
+            shareable = sharedContext == null ? helper.newShareable()
+                    : sharedContext.getOrCreate(new SharingKey(helper), helper::newShareable);
+        }
+
+        public <T extends FieldVector> T getVector(final Field field) {
+            try {
+                // noinspection unchecked
+                return (T) shareable.getReader().getVectorSchemaRoot().getVector(field);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
+
+        public void ensureLoadingBlock(final int blockNumber) {
+            this.shareable.ensureLoadingBlock(blockNumber);
+        }
+
+        @Override
+        public void close() {
+            if (!shared) {
+                helper.addToPool(shareable);
+            }
+        }
+    }
+
+    public static final class Shareable extends SharedContext {
```

Comment:
Can be a `ResettableContext` instead of a `SharedContext`. Also, see the changes Andy made in a round 2 commit.